### ISSUE TITLE
Redirect  to a custom search engine without Safari default ones

### ISSUE THREAD

**Issue by nyakojiru** [mshibanami/redirect-web#48]
How to redirect to a custom search engine Like SearXNG , without using one of the default search engines available in Safari? I don’t want to log my searches in any of them, it’s one of the main purpose I bought the app (privacy). Thank you.

**Comment by mshibanami** [comment]
@nyakojiru You can use the new DNR type to achieve your goal. This type doesn't initiate a request of fetching source web pages, as it uses [the DNR API](https://blog.mozilla.org/addons/2023/05/17/declarativenetrequest-available-in-firefox/), which is specially tailored for privacy by W3C.

For example, suppose you wish to use search.rhscz.eu, one of the SearXNG instances, as your default search engine. Follow these steps:

1. Set the default search engine to DuckDuckGo in **the Settings app** > **Safari** > **Search Engine**
2. Create the following redirect rule:
	* **Type**: DNR
	* **Redirect From**:
		* `https://duckduckgo.com/\?(?:.*&)?q=([^&]*).*` (Regular Expression)
		* **Resource Types**: `main_frame`
	* **Redirect To**: `https://search.rhscz.eu/search?q=$1`

I confirmed it didn't trigger a search query of DuckDuckGo, using Charles Proxy:

https://github.com/user-attachments/assets/fe980ff5-5c92-4569-8afc-c7b2935312b1

(`bag.itunes.apple.com` is an unrelated system request. Please ignore it.)

Optionally, you can disable **Search Engine Suggestions** from **the Settings app** > **Safari** so Safari won't even try to request search suggestions. Otherwise, it'll still call them like this:

https://github.com/user-attachments/assets/9521b4e6-f463-42aa-a231-3bc9ea145151

By the way, this is actually a great use case for user privacy that I didn't realize until now. I'll probably push this use case to other users too. Thanks a lot!

**Comment by mshibanami** [comment]
I'm closing this issue now. Please feel free to reopen it or create another issue if you have any othrer questions. You can also ask in https://github.com/mshibanami/redirect-web/discussions. Thanks.

**Comment by mshibanami** [comment]
@nyakojiru Oh that's weird. I double-checked it on my end but couldn't replicate it:

https://github.com/user-attachments/assets/8ad94738-0107-4cd3-beb3-2837f3300dc4

Would you please record a video of what it actually looks like? I'd especially like to see the following:

* The extension popup shown when tapping the icon on the toolbar
* The permissions you gave to the extension in Safari > Settings > Extensions
* The redirection rule you set up in the Redirect Web app
* The menu shown before pressing the enter key in the address bar

**Comment by mshibanami** [comment]
@nyakojiru Oh, that's strange. I double-checked it on my end but couldn't replicate it:

https://github.com/user-attachments/assets/bbac6321-164e-4338-a1b4-7b226a3b72e7

Could you please record a video showing what's happening on your side? Specifically, I'd like to see the following:

* The extension popup when tapping the icon on the toolbar
* The permissions you've granted to the extension in Safari > Settings > Extensions
* The redirection rule you've set up in the Redirect Web app
* The menu shown before pressing the enter key in the address bar

I hope this helps.

----

**HOWEVER**, I noticed that it no longer works on Safari Technology Preview 202. Safari doesn't load any web page when entering a query:

https://github.com/user-attachments/assets/52c6f6d4-bd21-49d8-b82b-cf38689b3cb2

I confirmed this issue also occurs on iOS 18.0 RC (22A3354) on my device.

It seems to have regressed, or perhaps it was intentional? I doubt it since it ruins the UX but it could be possible because they restrict the default search engines.
I'll ask about this on Apple's developer forum first, and depending on their response, I may raise the issue with [w3c/webextension](https://github.com/w3c/webextensions). Hope it's just degraded...

**Comment by mshibanami** [comment]
Replication:

- https://github.com/user-attachments/assets/e686ec6b-ee39-49e3-b0eb-9dcdb0cf6671
- Sample Xcode project: [DNRForDefaultSearchEngine.zip](https://github.com/user-attachments/files/16958822/DNRForDefaultSearchEngine.zip)

**Comment by mshibanami** [comment]
Posted: https://developer.apple.com/forums/thread/763505

**Comment by mshibanami** [comment]
Thanks for the information. As you can see, my post on the forums hasn't received any response from Apple yet. However, I believe they have already recognized the issue. Let's wait and see how it goes...

**Comment by mshibanami** [comment]
No, I don’t think so. If the DNR is a requirement for you, the control a third party has with DNR is extremely limited. (That makes DNR privacy oriented.)

You should either use the Original type or downgrade Safari.

**Comment by Falloutman** [comment]
Is there an update on this? I am having the same issue and had to disable the extension.

**Comment by mshibanami** [comment]
@Falloutman Unfortunately, I don't think so. Maybe Apple doesn't really care about this issue as of now.

Upvoting [my post](https://developer.apple.com/forums/thread/763505) that I previously mentioned could help catch Apple's attention.

Sorry, but this is the only thing I can suggest. The DNR API is intentionally limited so that browser extensions won't be able to manipulate or track user data, so the things I can do are very limited.
